lib/IRGen/GenProto.cpp

```diff
@@ -1710,32 +1839,78 @@ llvm::Constant *WitnessTableBuilder::buildInstantiationFunction() {
   if (IGM.DebugInfo)
     IGM.DebugInfo->emitArtificialFunction(IGF, fn);
 
+  auto PointerAlignment = IGM.getPointerAlignment();
+  auto PointerSize = IGM.getPointerSize();
+
   // Break out the parameters.
   Explosion params = IGF.collectParameters();
-  Address wtable(params.claimNext(), IGM.getPointerAlignment());
+  Address wtable(params.claimNext(), PointerAlignment);
   llvm::Value *metadata = params.claimNext();
-  llvm::Value *instantiationArgs = params.claimNext();
-  (void) instantiationArgs; // unused for now
-
-  // TODO: store any required conditional-conformance information
-  // in the private data.
+  Address instantiationArgs(params.claimNext(), PointerAlignment);
+
+  // The details of any conditional conformances are in the last argument, as a
+  // pointer/size pair.
+  auto conditionalTablesSlice = IGF.Builder.CreateElementBitCast(
+      instantiationArgs, IGF.IGM.WitnessTableSliceTy);
+
+  auto conditionalTablesPtr =
+      IGF.Builder.CreateStructGEP(conditionalTablesSlice, 0, Size(0));
+  Address conditionalTables(IGF.Builder.CreateLoad(conditionalTablesPtr),
+                            PointerAlignment);
+
+  auto numConditionalTablesPtr =
+    IGF.Builder.CreateStructGEP(conditionalTablesSlice, 1, Size(PointerSize));
+  auto numConditionalTables = IGF.Builder.CreateLoad(numConditionalTablesPtr);
+
+  // Assert that the number of witness tables passed in is the number required
+  // for the conditional conformances of this witness table.
+  auto failBB = IGF.createBasicBlock("bad_witness_table_count");
```

Comment:
I don't think we actually want to check this at runtime, although it might be useful for debugging for now

Reply:
Yes, I think it's good for debugging, especially while we bring up the dynamic casting.